packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -712,7 +716,88 @@ class _TheatreElement extends MultiChildRenderObjectElement {
   }
 }
 
-class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox, StackParentData> {
+// A `RenderBox` that sizes itself to its parent's size, implements the stack
+// layout algorithm and renders its children in the given `theatre`.
+mixin _RenderTheatreMixin on RenderBox {
+  _RenderTheatre get theatre;
+
+  RenderBox? get _firstOnstageChild;
+  RenderBox? get _lastOnstageChild;
+  int get _onstageChildCount;
+
+  @override
+  void setupParentData(RenderBox child) {
+    if (child.parentData is! StackParentData) {
+      child.parentData = StackParentData();
+    }
+  }
+
+  bool _layoutChild(RenderBox child, StackParentData childParentData, BoxConstraints nonPositionedChildConstraints, Alignment alignment) {
+    if (!childParentData.isPositioned) {
+      child.layout(nonPositionedChildConstraints, parentUsesSize: true);
+      childParentData.offset = alignment.alongOffset(size - child.size as Offset);
+      return false;
+    } else {
+      assert(child is! _RenderDeferredLayoutBox);
+      return RenderStack.layoutPositionedChild(child, childParentData, size, alignment);
+    }
+  }
+
+  @override
+  bool get sizedByParent => true;
+
+  @override
+  void performLayout() {
+    // `theatre` must override this method to reset _hasVisualOverflow.
+    RenderBox? child = _firstOnstageChild;
+    if (child == null) {
+      return;
+    }
+    // Same BoxConstraints as used by RenderStack for StackFit.expand.
+    final BoxConstraints nonPositionedChildConstraints = BoxConstraints.tight(constraints.biggest);
+    final Alignment alignment = theatre._resolvedAlignment;
+    while (child != null) {
+      final StackParentData childParentData = child.parentData! as StackParentData;
+      final bool hasVisualOverflow = _layoutChild(child, childParentData, nonPositionedChildConstraints, alignment);
+      theatre._hasVisualOverflow = theatre._hasVisualOverflow || hasVisualOverflow;
+      assert(child.parentData == childParentData);
+      child = childParentData.nextSibling;
+    }
+  }
+
+  @override
+  bool hitTestChildren(BoxHitTestResult result, { required Offset position }) {
+    RenderBox? child = _lastOnstageChild;
+    int childCount = _onstageChildCount;
+    while (child != null) {
+      final StackParentData childParentData = child.parentData! as StackParentData;
+      final bool isHit = result.addWithPaintOffset(
+        offset: childParentData.offset,
+        position: position,
+        hitTest: _toBoxHitTest(child.hitTest),
+      );
+      if (isHit) {
+        return true;
+      }
+      childCount -= 1;
+      child = childCount > 0 ? childParentData.previousSibling : null;
+    }
+    return false;
+  }
+
+  // `theatre` must override this method to handle clipping.
```

Comment:
uber-nit: unify the location of this comment? `performLayout` has it inside the method.